osu.Game/Rulesets/Objects/Types/ComboIndex.cs

```diff
@@ -0,0 +1,71 @@
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+
+namespace osu.Game.Rulesets.Objects.Types
+{
+    /// <summary>
+    /// Represents a manual and automated combo indices.
+    /// </summary>
+    public readonly struct ComboIndex : IEquatable<ComboIndex>
+    {
+        /// <summary>
+        /// The automated index value of the combo in relation to the beatmap.
+        /// </summary>
+        /// <remarks>
+        /// This is separated from <see cref="Manual"/> to disallow
+        /// manual combo colouring when not on the right skin to apply.
+        /// </remarks>
+        public readonly int Automated;
+
+        /// <summary>
+        /// The manual index value of the combo, this accounts for <see cref="IHasCombo.ComboOffset"/>.
+        /// </summary>
+        public readonly int Manual;
+
+        public ComboIndex(int automated, int manual)
+        {
+            Automated = automated;
+            Manual = manual;
+        }
+
+        public ComboIndex(int index)
+            : this(index, index)
+        {
+        }
+
+        public override string ToString() => $"(Automated: {Automated}, Manual: {Manual})";
+
+        #region Operator overloading
+
+        public static implicit operator ComboIndex(int i) => FromInt32(i);
+
+        public static bool operator ==(ComboIndex left, ComboIndex right) => left.Equals(right);
+        public static bool operator !=(ComboIndex left, ComboIndex right) => !(left == right);
+
+        public static ComboIndex operator +(ComboIndex left, ComboIndex right) => Add(left, right);
+        public static ComboIndex operator -(ComboIndex left, ComboIndex right) => Subtract(left, right);
+        public static ComboIndex operator *(ComboIndex left, ComboIndex right) => Multiply(left, right);
+        public static ComboIndex operator /(ComboIndex left, ComboIndex right) => Divide(left, right);
+
+        public static ComboIndex operator ++(ComboIndex ci) => Increment(ci);
+        public static ComboIndex operator --(ComboIndex ci) => Decrement(ci);
+
+        public static ComboIndex FromInt32(int i) => new ComboIndex(i);
+
+        public bool Equals(ComboIndex other) => Automated == other.Automated && Manual == other.Manual;
+        public override bool Equals(object obj) => obj is ComboIndex other && Equals(other);
+        public override int GetHashCode() => HashCode.Combine(Automated, Manual);
+
+        public static ComboIndex Add(ComboIndex a, ComboIndex b) => new ComboIndex(a.Automated + b.Automated, a.Manual + b.Manual);
+        public static ComboIndex Subtract(ComboIndex a, ComboIndex b) => new ComboIndex(a.Automated - b.Automated, a.Manual - b.Manual);
+        public static ComboIndex Multiply(ComboIndex a, ComboIndex b) => new ComboIndex(a.Automated * b.Automated, a.Manual * b.Manual);
+        public static ComboIndex Divide(ComboIndex a, ComboIndex b) => new ComboIndex(a.Automated / b.Automated, a.Manual / b.Manual);
```

Comment:
I don't think these make a lot of sense (why would adding two combo indices add them element-wise?). Do they need to exist? I don't think there is *any* sane way to define arithmetic operators for this type.

Reply:
I've added these such that the structure can almost act as a normal integer, such that custom implementations that need to interact / modify the combo index (say combo tests) can do so easily, for example:
```csharp
public void ApplyToHitCircle(HitCircle circle)
{
    Scheduler.AddDelayed(circle.ComboIndex += 2, 250);
}
```
since combo index is initialized as `(0, 0)`, adding 2 would do for both indices, and the combo index will become `(2, 2)` which will pick the third combo colour no matter the conditions. (as if you're adding 2 to an `int`)

Reply:
Yes, I get the idea. Thing is, with the current implementation, you also allow things like
```csharp
new ComboIndex(2, 6) + new ComboIndex(1, 5)
```
which will yield `new ComboIndex(3, 11)` but what I'm trying to get across is that *it doesn't really make much sense to do that semantically*. Those operations are ill-defined and *should not be possible to do in an implicit manner*. Even worse are the division operators. I don't think being able to do division on two instances of `ComboIndex` is going to ever be defensible.

Point is, (a) those are at this stage unused, and I generally discourage from trying to future-proof as it just produces dead code that might have unknown bugs, and (b) the API for manipulation of combo indices here is not really sane in many cases. Operator overloading in general I also think should be used *very* sparingly. At most I'd allow
```csharp
public struct ComboIndex
{
    public void Add(int amount);
    public void Subtract(int amount);
}
```
but even *then* strange issues can arise (what if you try to do `new ComboIndex(4, 0).Subtract(2)`? `ComboIndex(2, -2)` isn't really sane either, is it?)

Reply:
Went ahead with static `Add` and `Subtract` methods only with filled-up XMLDocs to avoid misusing it, I suppose that's the least sane way.